inspirehep/utils/date.py

```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2016 CERN.
+#
+# INSPIRE is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+"""Helpers for handling dates."""
+
+import six
+
+from dateutil.parser import parse
+
+from invenio_utils.date import datetime, strftime, strptime
+
+DATE_FORMATS_YEAR = ["%Y", "%y"]
+DATE_FORMATS_MONTH = [
+    "%Y-%m", "%Y %b", "%b %Y", "%Y %B", "%B %Y",
+    "%y-%m", "%y %b", "%b %y", "%y %B", "%B %y",
+]
+DATE_FORMATS_FULL = [
+    "%Y-%m-%d", "%d %m %Y", "%x", "%d %b %Y",
+    "%d %B %Y", "%d %b %y", "%d %B %y",
+]
+
+
+def create_valid_dates(dates):
+    """Add '01' as a value of missing month and/or day values."""
```

Comment:
The issue with adding `01` suffixes is that it introduces fake information that it's later trusted.

E.g. take this:
Say a document has dates: 2016 and 2016-05. What is its earliest date? With the proposed implementation is 2016-01-01, because of the fake `-01-01` being introduced. However it would very likely be more probable that the real earliest date is `2016-05-01`.  (2016 being just an unfortunate truncation due to how meta data is dealt with). I believe it's important to consider this info, otherwise it will look like that all papers receive citations only at the first of January of every year.

In order to achieve this, on legacy, we were not touching the actual metadata, but for when it was really necessary: 2016 was stored as 2016, and only at runtime, when the question "earliest_date" was asked, it was converted temporary to:

`2016-99-99` so that then it could have been compared with e.g. `2016-05-99`. Then the earliest date is clearly `2016-05-99`, which contains this canary value `-01` that can be converted at the end to `-01` giving the desired `2016-05-01`.